### DISCUSSION TITLE
Beckhoff ADS

### DISCUSSION THREAD

**Discussion by makelju** [frangoteam/FUXA#380] [Ideas]
Hello!
This SCADA looks great and is simple to use.

Would it be possible to add Beckhoff ADS communication as one connection option?

We mainly use Beckhoff PLC in our work projects. And this SCADA would be a good fit for the smaller projects we’re doing.

**Comment by unocelli** [comment]
Hi,
I have seen that there is a [library](https://github.com/roccomuso/node-ads) that we could try to integrate. but unfortunately I don't use and am not in possession of a Beckhoff PLC to test it. :confused:

**Reply by makelju** [reply]
Hi,
All Beckhoff PLCs are just windows PCs where the PLC program runs under TwincatXAE program. 
TwinCAT engineering software (TE1000) is free to download from Beckhoff and when installing this engineering software it will install TwinCatXAE-runtime as well.
Then you can run Beckhoff PLC-code in your computer same way as with a real PLC.

Beckhoff also has its own javascrip library for ADS communication but i dont know if this works with Nodejs.
This library is in Twincat installation folder. (Twincat\AdsApi\TcAdsWebService\AdsJavascriptLibrary\)
There is also other ADS libraries for other programming languages.

**Reply by unocelli** [reply]
Thanks. I should learn how to use TwinCAT and write Beckhoff PLC-code...In don't have resource for this :worried:

**Reply by henjoe** [reply]
I believed Beckoff PLCs have builtin opcua server, I'd radther use it than creating additional plugin in Fuxa.

**Comment by cyrilrey** [comment]
Hi, Fuxa is very promising and I also will be really interested if it allow ads native Beckhoff communication. Beckhoff is widespread in building automation this will be a good push for the popularity of Fuxa.

@unocelli, I understood that setting up an dev environment as suggested by @makelju is too demanding.

Else I guess, may it be possible to setup a preliminary test environment with : https://www.npmjs.com/package/ads-server.

    Use case of the project mentioned : "Creating a "fake PLC" to test your system"

Maybe @jisotalo may also provide suitable ready to use code and I guess he also may be happy to see his library pushed into Fuxa?
So, I would also suggest using @jisotalo related ads client for fuxa : https://github.com/jisotalo/ads-client. That library seems to be well made.

To be confirmed, but, I guess the communication mechanics in Fuxa already exist with Bacnet protocol and OPC UA and may be extended for Ads ? @unocelli, could you evaluate the scope of the work ? Is is a big deal or not ?

@unocelli I see that you are located in Bern. I am in Valais, so not fare away. I am not javascript programmer, but may help as Beckhoff integrator and programmer. I think we may also ask Beckhoff office in Yverdon to lend devices if needed.

**Reply by unocelli** [reply]
Nice to meet you, yes the library looks very interesting. i would say it is more a question of priority than feasibility.

**Reply by cyrilrey** [reply]
Nice to meet you too. I understood what you mean about priority management.

We can keep things open and if someone motivated and with javascript skills comes along, I think, we can do something together.

**Comment by jisotalo** [comment]
Hi all!

The https://github.com/jisotalo/ads-client library would certainly do the job. It shouldn't be too hard to be implemented in FUXA. I'm glad it was mentioned here!

Actually we have had some email discussion with @unocelli and he is working on it as far as I know.

**Comment by unocelli** [comment]
Hi,
Yes, the work is in progress in the `devices` branch. @jisotalo Thanks for your support!

**Comment by unocelli** [comment]
Hi, I update the `devices` branch with target/local/router address but I can't test it, is anyone interested?

**Reply by 99cm234** [reply]
did this ever get tested?

**Comment by sakloui** [comment]
Hey all, I've created a pull request [here](https://github.com/frangoteam/FUXA/pull/1661) which updates @unocelli 's devices branch to the latest version (and also ads-client), if anyone is still interested you can try it out.

I will be working on it for the next couple weekends, and I hope to get it merged sometime soon.

**Reply by henjoe** [reply]
Looking forward to this, and I am willing to test it (though I dont have any beckoff PLC 😅🤷)

**Reply by sakloui** [reply]
Neither do I at home 😅, at work I have a couple I can play around with.
Right now my setup is a Windows 10 VM for TwinCAT 3, a TC/BSD VM and then my host PC and all share the same isolated network so I don't run into any routing issues.

**Reply by cmarinx** [reply]
@sakloui 
Hi, Today I tried your /feature-ads.  I don,t know the reason, but I am not able to create a new connection - ADSclient. On Device property  select menu, is not listed as entry. Can you give me a Hint? how to fix it?

**Reply by sakloui** [reply]
Heya @cmarinx, 
I assume you directly ran the server with ```npm run start```
I purposefully did not include the built client code, so you will have to built it yourself, which you can do by running ```npm run build``` in the client directory. This should make the ADSclient show in the drop down menu.

I've unfortunately not had the time to update the branch, but it should be working well enough for testing and playing around in office. I still have plans to work on it further, but the lack of support from the company I work at and busy weekends slows it down quite a bit.

Edit:

Also if you're running this on TwinCat BSD/Linux make sure to open port 48898, or try to use the secure port 8016 (I have not tried this port and I doubt it works).

Some images of my settings and the received data below:
<details><summary>Configured settings</summary>
<img src="https://github.com/user-attachments/assets/b9796514-d7d0-416e-b2c9-56043f0f47c4">
<img src="https://github.com/user-attachments/assets/006cf2a7-1215-4c02-ba95-0e4fdab92e4f">
<img src="https://github.com/user-attachments/assets/3e0c34ea-7e4a-45a5-87aa-b56ece0e80ff">
</details>

**Reply by cmarinx** [reply]
Hi @sakloui ,
Thank you very much! 

indeed I did not build it. After I bulid it is working, I can read values from Twincat, But  writing walues does not work. When I try to write values, values in PLC are nulled. This is  the actual state, or also hier is something to be set up before use?
Do you plan to extend data type to Lreal  String?

Reading PLC Values,.

![readPLCValues](https://github.com/user-attachments/assets/cbca3a2b-2244-4758-8dfd-fdd61d14de7f)

Test write values:

![image](https://github.com/user-attachments/assets/6201c492-f823-4656-a39e-a2e9a1ddf335)

*------

![image](https://github.com/user-attachments/assets/97265c72-b2cc-4874-90ed-23975816ba2b)

**Reply by sakloui** [reply]
@cmarinx  The reason why it was writing 0 to the controller was due to the fact that I trying to write a promise object to the controller :sweat_smile:. I've fixed it now, and you should now be able to write to it.
I also changed the type selection to be the correct types (and it also stores it now), and brought the branch back up to date with the master branch.

<details><summary>Images</summary>
<p>

![image](https://github.com/user-attachments/assets/3ad039c3-dee1-4288-8f5f-adc74dcdbce7)
![image](https://github.com/user-attachments/assets/e34f19c7-0d22-422a-8f90-9ce9d4ad27ed)

</p>
</details>

**Reply by cmarinx** [reply]
@sakloui nice Job. One little observation. When you try to remove Device, the buttons are not  displayed anymore.

![image](https://github.com/user-attachments/assets/8e21950f-cbce-4a40-8374-a15fe1b31420)

**Reply by sakloui** [reply]
> @sakloui nice Job. One little observation. When you try to remove Device, the buttons are not displayed anymore.
> 

It should be fixed now @cmarinx, have you been ran into any other issues? If not I'm going to see if it can be merged into main for initial implementation.

**Comment by cmarinx** [comment]
@sakloui until now I did not have time to test it, just did it in small chunks, only in my free time. In weekend I plan to write a test it with  more than 500 Variable. I've found this observation on Infosys. 

![image](https://github.com/user-attachments/assets/ff7632a8-32ab-478d-a05e-b83a5a15638f)

**Reply by sakloui** [reply]
@cmarinx That sounds good! Let me know how well it handles it. @unocelli merged the PR into main, so you can use that as well now :)

**Comment by unocelli** [comment]
@sakloui Merge complete – Added dynamic handling when ads-client is not installed.
The system now checks if the ads-client package is available and adjusts accordingly, avoiding errors in setups where it's not needed.
Great job on this! Let’s keep improving as we go.

**Reply by sakloui** [reply]
Perfect! Thanks for all the help.
We can still improve it further, I wanted to see if I can take a shot at browsing.
